File: credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Ограничение на количество цифр в карте [от 1 до 16 цифр]
    long long card;
    do
    {
        card = get_long_long("Number:\n");
    }
    while (card < 1 || card > 9999999999999999);

    // Создаем переменные счетчики
    int count = 0;
    int sum1 = 0;
    int sum2 = 0;
    int ls_dig = 0;
    int ls_dig1 = 0;
    int ls_dig2 = 0;
    long long copy_card = card;
    long long copy_card2 = card;
    long long company_id = card;

    // Количество цифр в номере карты
    while (card > 0)
    {
        ls_dig = card % 10;
        count++;
        card = card / 10;
    }
    // Сумма вторых цифр карты (с права на лево)
    for (int j = 0; j < count / 2; j++)
    {
        ls_dig1 = copy_card % 100 / 10 * 2;
        if (ls_dig1 > 9)
        {
            ls_dig1 = ls_dig1 / 10 + ls_dig1 % 10;
        }
        sum1 = sum1 + ls_dig1;
        copy_card = copy_card / 100;
    }
    // Сумма первых цифр карты (с права на лево)
    for (int i = 0; i < count / 2 + 1; i++)
    {
        ls_dig2 = copy_card2 % 10;
        sum2 = sum2 + ls_dig2;
        copy_card2 = copy_card2 / 100;
    }
    int sum = sum1 + sum2;
    // Первые две цифры карты
    while (company_id >= 99)
    {
        company_id /= 10;
    }

    // Определение платежной системы
    if (count == 16 && (51 <= company_id && company_id <= 55) && (sum % 10 == 0))
    {
        printf("MASTERCARD\n");
    }
    else if ((count == 13 || count == 16) && (company_id / 10 == 4) && (sum % 10 == 0))
    {
        printf("VISA\n");
    }
    else if (count == 15 && (company_id == 34 || company_id == 37) && (sum % 10 == 0))
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }

}

File: Week 6/credit.py
from cs50 import get_int
import re

# Check count of digit in card
card = get_int("Number: ")
while card < 1 or card > 9999999999999999:
    card = get_int("Number: ")

# Count digit in card
count = len(str(card))

# Decimal conversation


def get_dec_num(n):
    return n // 10 + n % 10
    

# Luhn’s Algorithm
copy_card = card
sum1, sum2 = 0, 0

while copy_card > 0:
    l_dig1 = copy_card % 100 // 10 * 2
    l_dig2 = copy_card % 10
    if l_dig1 > 9:
        l_dig1 = get_dec_num(l_dig1)
    sum1 += l_dig1
    sum2 += l_dig2
    copy_card = copy_card // 100
sum = sum1 + sum2

# First two digits of the card
company_id = str(card)[:2]

# Card verification
if str(sum).endswith("0"):
    if count == 16 and (51 <= int(company_id) <= 55):
        print("MASTERCARD")
    elif count in [13, 16] and re.search("^4", company_id):
        print("VISA")
    elif count == 15 and re.findall("34|37", company_id):
        print("AMEX")
else:
    print("INVALID")